newrelic/metric.go

```diff
@@ -0,0 +1,226 @@
+	Metrics         []data   `json:"metrics`
+}
+type responseData struct {
+	MetricData metricData `json:"metric_data"`
+}
+
+// Client defines the information necessary to query New Relic
+type Client struct {
+	APIKey        string
+	ApplicationId string
+	BaseUrl       string
+}
+
+// Query returns a unique name for the query
+func (m *Metric) Query() string {
+	return fmt.Sprintf("%v:%v", m.config.MetricName, m.config.MetricValue)
+}
+
+// QueryMetrics returns a metric descriptor and a slice of TimeSeries for the metric.
+// all elements of the TimeSeries must be the same metric descriptor and contain a single point
+func (m *Metric) QueryMetrics(ctx context.Context, start time.Time, end time.Time, name string, value string) (*metricpb.MetricDescriptor, []*monitoringpb.TimeSeries, error) {
+	c := m.client
+
+	var ret []*monitoringpb.TimeSeries
+	tmpStart := start
+
+	for continueLoop := true; continueLoop; {
+		// Always use 3 hour windows, anything more and points will get aggregated to greater than 1 minute granularity
+		tmpEnd := tmpStart.Add(3 * time.Hour)
+		if tmpEnd.After(end) {
+			tmpEnd = end
+			continueLoop = false
+		}
+
+		// TODO: add multi-page support
+		url := fmt.Sprintf("%s/v2/applications/%v/metrics/data.json",
+			c.BaseUrl, c.ApplicationId)
+		req, err := http.NewRequest("GET", url, nil)
+		if err != nil {
+			return nil, nil, err
+		}
+		q := req.URL.Query()
+		q.Add("names[]", name)
+		q.Add("values[]", value)
+		q.Add("from", tmpStart.Format(time.RFC3339))
+		q.Add("to", tmpEnd.Format(time.RFC3339))
+
+		req.URL.RawQuery = q.Encode()
+		req.Header.Set("X-Api-Key", c.APIKey)
+		httpClient := &http.Client{}
+		resp, err := httpClient.Do(req)
+		if err != nil {
+			return nil, nil, err
+		}
+		defer resp.Body.Close()
+		body, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		var bodydata responseData
+		err = json.Unmarshal(body, &bodydata)
+		if err != nil {
+			return nil, nil, err
+		}
+		ts := m.convertTimeSeries(ctx, bodydata, value)
+
+		if ret == nil || len(ret) == 0 {
+			ret = ts
+		} else {
+			if len(ts) > 0 {
+				ret[0].Points = append(ret[0].Points, ts[0].Points...)
+			}
+		}
+
+		tmpStart = tmpEnd
+	}
+	descriptor := &metricpb.MetricDescriptor{
+		Name:        m.StackdriverName(),
+		MetricKind:  metricpb.MetricDescriptor_GAUGE,
+		ValueType:   metricpb.MetricDescriptor_DOUBLE,
+		Description: fmt.Sprintf("New Relic Query for name: %v, value %v\n", name, value),
+		DisplayName: fmt.Sprintf("%v:%v", name, value),
+		Type:        m.StackdriverName(),
+	}
+	return descriptor, ret, nil
+}
+
+func (m *Metric) convertTimeSeries(ctx context.Context, data responseData, value string) []*monitoringpb.TimeSeries {
+	var ts []*monitoringpb.TimeSeries
+
+	timeslices := data.MetricData.Metrics[0].Timeslices
+	if len(timeslices) == 0 {
+		return nil
+	}
+
+	for _, slice := range timeslices {
+		val, ok := slice.Values[value]
```

Comment:
Could you please check that `slice.Values` does not contain multiple values, and return an error if it does?

Reply:
This comment does not seem to be addressed.